apps/server/src/modules/tldraw/utils/utils.ts

```diff
@@ -0,0 +1,254 @@
+			encoding.writeVarUint(encoder, WSMessageType.AWARENESS);
+			encoding.writeVarUint8Array(encoder, encodeAwarenessUpdate(this.awareness, changedClients));
+			const buff = encoding.toUint8Array(encoder);
+			this.conns.forEach((_, c) => {
+				send(this, c, buff);
+			});
+		};
+		this.awareness.on('update', awarenessChangeHandler);
+		this.on('update', updateHandler);
+		if (this.isCallbackSet) {
+			this.on('update', debounce(callbackHandler, CALLBACK_DEBOUNCE_WAIT, { maxWait: CALLBACK_DEBOUNCE_MAX_WAIT }));
+		}
+	}
+}
+
+/**
+ * Gets a Y.Doc by name, whether in memory or on disk
+ *
+ * @param {string} docname - the name of the Y.Doc to find or create
+ * @param  gc - whether to allow gc on the doc (applies only when created)
+ * @return {WSSharedDoc}
+ */
+export const getYDoc = (docname: string, gc = true) =>
+	map.setIfUndefined(docs, docname, () => {
+		const doc = new WSSharedDoc(docname);
+		doc.gc = gc;
+		if (persistence !== null) {
+			void persistence.bindState(docname, doc);
+		}
+		docs.set(docname, doc);
+		return doc;
+	});
+
+/**
+ * @param {WebSocket} conn
+ * @param {WSSharedDoc} doc
+ * @param {Uint8Array} message
+ */
+export const messageHandler = (conn: WebSocket, doc: WSSharedDoc, message: Uint8Array) => {
+	try {
+		const encoder = encoding.createEncoder();
+		const decoder = decoding.createDecoder(message);
+		const messageType = decoding.readVarUint(decoder);
+		switch (messageType) {
+			case WSMessageType.SYNC:
+				encoding.writeVarUint(encoder, WSMessageType.SYNC);
+				readSyncMessage(decoder, encoder, doc, conn);
+
+				// If the `encoder` only contains the type of reply message and no
+				// message, there is no need to send the message. When `encoder` only
+				// contains the type of reply, its length is 1.
+				if (encoding.length(encoder) > 1) {
+					send(doc, conn, encoding.toUint8Array(encoder));
+				}
+				break;
+			case WSMessageType.AWARENESS: {
+				applyAwarenessUpdate(doc.awareness, decoding.readVarUint8Array(decoder), conn);
+				break;
+			}
+			default:
+				break;
+		}
+	} catch (err) {
+		console.error(err);
+		doc.emit('error', [err]);
+	}
+};
+
+/**
+ * @param {WebSocket} ws
+ * @param {string} docName
+ */
+export const setupWSConnection = (ws: WebSocket, docName = 'GLOBAL') => {
+	ws.binaryType = 'arraybuffer';
+	// get doc, initialize if it does not exist yet
+	const doc = getYDoc(docName, true);
+	doc.conns.set(ws, new Set());
+
+	// listen and reply to events
+	ws.on('message', (message: ArrayBufferLike) => {
+		messageHandler(ws, doc, new Uint8Array(message));
+	});
+
+	// Check if connection is still alive
+	let pongReceived = true;
+	const pingInterval = setInterval(() => {
+		if (!pongReceived) {
+			if (doc.conns.has(ws)) {
+				closeConn(doc, ws);
+			}
+			clearInterval(pingInterval);
+		} else if (doc.conns.has(ws)) {
+			pongReceived = false;
+			try {
+				ws.ping();
+			} catch (e) {
+				closeConn(doc, ws);
+				clearInterval(pingInterval);
+			}
+		}
```

Comment:
Please check if you could reduce the complexity here a bit - it's a little hard to understand all the possible conditions, maybe we can do some faster returns in some cases to make the flow end a bit sooner in some cases?

Reply:
Maybe something like that?

```ts
const hasConn = doc.conns.has(ws);

if (pongReceived) {
	if (!hasConn) return;

	pongReceived = false;

	try {
		ws.ping();
	} catch (e) {
		closeConn(doc, ws);
		clearInterval(pingInterval);
	}

	return;
}

if (hasConn) {
	closeConn(doc, ws);
}

clearInterval(pingInterval);
```